pkg/engine/datasource/kafka_datasource/sarama_config_parameters_test.go

```diff
@@ -0,0 +1,446 @@
+
+	retryFn := func() error {
+		switch conn.State() {
+		case zk.StateHasSession, zk.StateConnected:
+			return nil
+		default:
+			return errors.New("not yet connected")
+		}
+	}
+
+	require.NoError(t, k.pool.Retry(retryFn))
+	t.Log("ZooKeeper has been started")
+}
+
+func (k *kafkaBroker) startKafka(t *testing.T) *dockertest.Resource {
+	t.Log("Trying to run Kafka")
+
+	var envVars []string
+	for _, cfg := range basicKafkaEnvVars {
+		envVars = append(envVars, cfg)
+	}
+	for _, cfg := range k.config.kafkaEnvVars {
+		envVars = append(envVars, cfg)
+	}
+
+	resource, err := k.pool.RunWithOptions(&dockertest.RunOptions{
+		Name:       "kafka-tyk-graphql",
+		Repository: "bitnami/kafka",
+		Tag:        "3.0.1",
+		NetworkID:  k.network.ID,
+		Hostname:   "kafka",
+		Env:        envVars,
+		PortBindings: map[docker.Port][]docker.PortBinding{
+			"9092/tcp": {{HostIP: "localhost", HostPort: "9092/tcp"}},
+		},
+		ExposedPorts: []string{"9092/tcp"},
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, k.pool.Purge(resource))
+	})
+
+	retryFn := func() error {
+		config := sarama.NewConfig()
+		config.Producer.Return.Successes = true
+		config.Producer.Return.Errors = true
+		brokerAddr := fmt.Sprintf("localhost:%s", resource.GetPort("9092/tcp"))
+		asyncProducer, err := sarama.NewAsyncProducer([]string{brokerAddr}, config)
+		if err != nil {
+			return err
+		}
+		defer asyncProducer.Close()
+
+		var total int
+	loop:
+		for {
+			total++
+			if total > 100 {
+				break
+			}
+			message := &sarama.ProducerMessage{
+				Topic: "grahpql-go-tools-health-check",
+				Value: sarama.StringEncoder("hello, world!"),
+			}
+
+			asyncProducer.Input() <- message
+
+			select {
+			case <-asyncProducer.Errors():
+				// We should try again
+				//
+				// Possible error msg: kafka: Failed to produce message to topic grahpql-go-tools-health-check:
+				// kafka server: In the middle of a leadership election, there is currently no leader for this
+				// partition and hence it is unavailable for writes.
+				continue loop
+			case <-time.After(time.Second):
+				continue loop
+			case <-asyncProducer.Successes():
+				break loop
+			}
+
+		}
+		return nil
+	}
+
+	if err = k.pool.Retry(retryFn); err != nil {
+		log.Fatalf("could not connect to kafka: %s", err)
+	}
+	require.NoError(t, k.pool.Retry(retryFn))
+
+	t.Log("Kafka is ready to accept connections")
+	return resource
+}
+
+func (k *kafkaBroker) start(t *testing.T) *dockertest.Resource {
+	t.Cleanup(func() {
+		require.NoError(t, k.pool.Client.RemoveNetwork(k.network.ID))
+	})
+	k.startZooKeeper(t)
+	return k.startKafka(t)
+}
+
+func testAsyncProducer(t *testing.T, addr, topic string, start, end int) {
+	config := sarama.NewConfig()
+	asyncProducer, err := sarama.NewAsyncProducer([]string{addr}, config)
+	require.NoError(t, err)
+
+	for i := start; i < end; i++ {
+		message := &sarama.ProducerMessage{
+			Topic: topic,
+			Value: sarama.StringEncoder(fmt.Sprintf("message-%d", i)),
```

Comment:
this issue was raised by SonarCloud: Maybe it makes sense to create a constant for it because it is used several times (`"message-%d"`)